File: CSharp/demo-Search/Search.Contracts/Models/FilterExpression.cs
namespace Search.Models
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    public enum Operator { None, LessThan, LessThanOrEqual, Equal, GreaterThanOrEqual, GreaterThan, And, Or, Not, FullText };

    [Serializable]
    public class FilterExpression
    {
        public Operator Operator { get; }
        public object[] Values { get; }
        public string Description { get; }

        public FilterExpression(string description, Operator op, params object[] values)
        {
            Description = description;
            Operator = op;
            Values = values;
        }

        public FilterExpression(Operator op, params object[] values) : this(null, op, values) { }

        public override string ToString()
        {
            var builder = new StringBuilder();
            builder.Append(Operator.ToString());
            builder.Append("(");
            var seperator = "";
            foreach (var value in Values)
            {
                builder.Append(seperator);
                builder.Append(value.ToString());
                seperator = ", ";
            }
            builder.Append(")");
            return builder.ToString();
        }

        /// <summary>
        /// Obtains a user-friendly description of the filter.
        /// Traverses the expression tree collecting the descriptions from the nodes.
        /// </summary>
        public string ToUserFriendlyString()
        {
            const string spacePrefix = " ";
            var prefix = string.Empty;
            var builder = new StringBuilder();

            VisitExpressionTree(this, expression =>
            {
                //If the expression does not have a description, a descendant should.
                if (string.IsNullOrEmpty(expression.Description))
                    return true;

                builder.Append($"{prefix}\"{expression.Description}\"");

                if (prefix == string.Empty)
                {
                    prefix = spacePrefix;
                }
                //Since we found a description, descendants, if any, should not have a description. We stop traversing this branch.
                return false;
            });
            return builder.ToString();
        }

        /// <summary>
        /// Visits the nodes of the expression tree, and stops recursing a given branch when the visitor returns false.
        /// </summary>
        private static void VisitExpressionTree(FilterExpression node, Func<FilterExpression, bool> expressionVisitor)
        {
            if (node == null) return;

            //We execute the visitor function and keep recursing this branch or not depending on the visitor's result
            bool shouldKeepVisiting = expressionVisitor(node);

            if (shouldKeepVisiting)
            {
                foreach (var value in node.Values)
                {
                    var filter = value as FilterExpression;
                    if (filter != null)
                    {
                        VisitExpressionTree(filter, expressionVisitor);
                    }
                }
            }
        }

        public FilterExpression DeepCopy()
        {
            var values = (from value in Values select value is FilterExpression ? (object)(value as FilterExpression).DeepCopy() : value).ToArray();
            return new FilterExpression(Description, Operator, values);
        }

        // Remove all references to the same field
        public FilterExpression Remove(FilterExpression expression)
        {
            FilterExpression filter = this;
            if (expression != null)
            {
                foreach (var field in expression.Fields())
                {
                    filter = filter.Remove(field);
                }
            }
            return filter;
        }

        public IEnumerable<SearchField> Fields()
        {
            return AllFields().Distinct();
        }

        private IEnumerable<SearchField> AllFields()
        {
            if (Operator == Operator.And
                || Operator == Operator.Or
                || Operator == Operator.Not)
            {
                foreach(FilterExpression child in Values)
                {
                    foreach(var field in child.Fields())
                    {
                        yield return field;
                    }
                }
            }
            else 
            {
                foreach(var value in Values)
                {
                    if (value is SearchField)
                    {
                        yield return value as SearchField;
                    }
                }
            }
        }

        public FilterExpression Remove(SearchField field)
        {
            FilterExpression result = null;
            if (Operator == Operator.And
                || Operator == Operator.Or)
            {
                var child1 = (Values[0] as FilterExpression).Remove(field);
                var child2 = (Values[1] as FilterExpression).Remove(field);
                result = FilterExpression.Combine(child1, child2, Operator, Description);
            }
            else if (Operator == Operator.Not)
            {
                var child = (Values[0] as FilterExpression).Remove(field);
                if (child != null)
                {
                    result = new FilterExpression(Description, Operator.Not, child);
                }
            }
            else
            {
                if (Values.All((v) => v is SearchField ? (v as SearchField).Name != field.Name : true))
                {
                    result = this;
                }
            }
            return result;
        }

        public static FilterExpression Combine(FilterExpression child1, FilterExpression child2, Operator combination = Operator.And, string description = null)
        {
            FilterExpression filter;
            if (child1 != null)
            {
                if (child2 != null)
                {
                    filter = new FilterExpression(description, combination, child1, child2);
                }
                else
                {
                    filter = child1;
                }
            }
            else
            {
                filter = child2;
            }
            return filter;
        }
    }
}

File: Node/demo-Search/src/model/FilterExpression.ts
// 
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license.
// 
// Microsoft Bot Framework: http://botframework.com
// 
// Bot Builder SDK Github:
// https://github.com/Microsoft/BotBuilder
// 
// Copyright (c) Microsoft Corporation
// All rights reserved.
// 
// MIT License:
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
// 
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED ""AS IS"", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import { ISearchField } from '../search/ISearchSchema';
import { StringBuilder } from '../tools/StringBuilder';
import { sprintf } from 'sprintf-js';

export enum Operator {
    None, LessThan, LessThanOrEqual, Equal, GreaterThanOrEqual, GreaterThan, And, Or, Not, FullText
}

export class FilterExpression {

    private readonly values: any[];

    constructor(private description: string, private operator: Operator,  ...args:any[]) {
        this.values = args;
    }

    public static combine(left: FilterExpression, right: FilterExpression, operator: Operator = Operator.And, description: string = null): FilterExpression {
        let expression: FilterExpression;

        if(left && right) {
            expression = new FilterExpression(description, operator, left, right);
        } else if (left) {
            expression = left;
        } else if (right) {
            expression = right;
        }
        return expression;
    }

    public remove(expression: FilterExpression): FilterExpression {

        let filter: FilterExpression = this;

        if(expression) {
            for(let field of expression.retrieveSearchFields()) {
                if(filter) {
                    filter = filter.removeSearchField(field);
                }
            }
        }

        return filter;
    }

    public retrieveSearchFields(): ISearchField[] {
        
        let fields = new Array<ISearchField>(); 

        // Recursively retrieve all distinct search fields from the tree.
        FilterExpression.traversePreOrder(this, (node: FilterExpression): boolean => {

            if(node.operator == Operator.And || node.operator == Operator.Or || node.operator == Operator.Not) {
                return true;
            }
            else {
                for(let value of node.values) {
                    if(this.isSearchField(value)) {
                        fields.push(value);
                    }
                }
                return false;
            }
        });

        //Rudimentary way of removing duplicates
        return fields.filter((value: ISearchField, index: number, array: ISearchField[]): any => {
             return array.lastIndexOf(value) == index
         });
    }

    public removeSearchField(fieldToRemove: ISearchField): FilterExpression {
        
        let resultExpression: FilterExpression = null; 

        // Recursively remove all occurrences of the specified search field from the expression tree.
        FilterExpression.traversePostOrder<FilterExpression>(this, (node: FilterExpression, childrenResults: FilterExpression[]): FilterExpression => {

            if(node.operator == Operator.And || node.operator == Operator.Or) {
                resultExpression = FilterExpression.combine(childrenResults[0], childrenResults[1], node.operator, node.description);
            }
            else if (node.operator == Operator.Not) {
                if(childrenResults.length == 1 && childrenResults[0]) {
                    resultExpression = new FilterExpression(node.description, Operator.Not, childrenResults[0]);
                }
            }
            else {
                if(node.values.every((value: any, index: number, array: any[]): boolean => {
                    if(this.isSearchField(value)) {
                        if((<ISearchField>value).Name == fieldToRemove.Name) {
                            return false;
                        }
                    }
                    return true;
                }, this)) {
                    resultExpression = node;
                }
            }

            return resultExpression;
        });
        return resultExpression;
    }

    public toUserFriendlyString(): string {
        const spacePrefix = ' ';
        const emptyPrefix = '';
        const expressionFormat = '%s\"%s\"';
        let prefix = emptyPrefix;
        let stringBuilder = new StringBuilder();
        
        FilterExpression.traversePreOrder(this, (node: FilterExpression): boolean => {
            if(!node.description) {
                return true;
            } 

            stringBuilder.append(sprintf(expressionFormat, prefix, node.description));

            if(prefix == emptyPrefix) {
                prefix = spacePrefix;
            }

            return false;
        });

        return stringBuilder.toString();
    }

    private isSearchField(value: any): boolean {
        return value.Name && value.Type;
    }

    public static traversePreOrder(node: FilterExpression, nodeVisitor: (node: FilterExpression) => boolean): void {
        if(!node) return;

        // We execute our visitor function and keep recursing this branch of the tree or not depending on the visitor's result
        let shouldKeepVisiting: boolean = nodeVisitor(node);

        if(shouldKeepVisiting) {
            for(let value of node.values) {
                if(value instanceof FilterExpression) {
                    FilterExpression.traversePreOrder(<FilterExpression>value, nodeVisitor);
                }
            }
        }
    }

    public static traversePostOrder<TResult>(node: FilterExpression, nodeVisitor: (node: FilterExpression, childrenResults: TResult[]) => TResult): TResult {

        let results = new Array<TResult>();

        for(let value of node.values) {
            if(value instanceof FilterExpression) {
                results.push(FilterExpression.traversePostOrder<TResult>(<FilterExpression>value, nodeVisitor));
            }
        }

        return nodeVisitor(node, results);
    }

    public getOperator(): Operator {
        return this.operator;
    } 
    public getValues(): any[] {
        return this.values;
    }
}